New class `Parameter`. Added names for DNNLayer. Update optimizer. [WIP]

KOHbDS opened the pull request:
A new `Parameter` class has been created. Added names for `DNNLayer`. Names are included in the `optimizer.py` for `GradientDescent`. The necessary changes to the 'test_neural_net.py' have been made.

haydnv commented on client/tinychain/ml/optimizer.py:
```diff
@@ -16,26 +17,44 @@ def optimize(self, i, gradient, inputs, loss):
-                return [update(gradient[n], deltas[n]) for n in range(len(shape))]
-            else:
-                return gradient.write(gradient - (deltas * lr))
+        self.m = [m.value.write(m.value * beta1 + p.grad * (F32(1) - beta1)) for m, p in zip(self.m, param_list)]
```
These layer names are implementation details of a `DNN` which other `Differentiable`s, like a `CNN`, won't share. It would be better to do automatic discovery of the parameters using the `shape` of the model to train. I think that would also remove the need for the `name` parameter in `DNNLayer.load`.

KOHbDS replied:
Done.
It was a draft version.

KOHbDS commented:
In the `GradientDescent` tests, I changed the number of layers and the test failed. Previously layers were with shape [n, n] and I changed to [m, n] (n != m). I will look at the transfer between the layers in more detail. It would be great if you also watched, maybe I missed something.

haydnv commented on client/tinychain/ml/__init__.py:
```diff
@@ -98,3 +99,11 @@ def create(cls, *args, **kwargs):
+
+
+class Parameter:
+
```
I think it would be better to use `Tensor` here, or just not specify a type (that way this will still work if in the future someone decides to use a `Sparse` tensor)

haydnv commented on client/tinychain/ml/dnn.py:
```diff
@@ -3,21 +3,22 @@
 class DNNLayer(Layer):
     @classmethod
-    def create(cls, input_size, output_size, activation=Sigmoid()):
+    def create(cls, name: str, input_size, output_size, activation=Sigmoid()):
```
This breaks the convention where a TinyChain `State` doesn't know it's own name, which is important because TinyChain has no global namespace to check for potential collisions. For example, this would work fine:

```python
def create_layer():
    return DNNLayer("layer", 10, 10)
```

But then this would break:
```python
layers = [create_layer() for _ in range(num_layers)]
```

You could add a "layer_num" parameter to `create_layer` but that still assumes that the calling namespace has only one DNN, etc. If you need a unique identifier at compile time, you can use the `hex_id` function in `tinychain.util`.

haydnv replied:
Actually, I thought about this more. I like the idea of having descriptive names for debugging, we just need to make sure that the top-level calling code (e.g. an Optimizer) can provide a top-level namespace. For example, when an Optimizer is created for a model called "my_nn" it can initialize parameters like "my_nn.0.weights", "my_nn.0.bias" etc. That way if a user uses two instances of my_nn in a larger differentiable module they can be assigned names like "larger_module.my_nn..." and there won't be a collision.

KOHbDS replied:
The name collision is impossible when the following requirements are fulfilled:
1. The calling code guarantees the unique names of the layers.
2. Every optimizer instance is serving a single model exclusively.
This model/optimizer usage pattern is quite typical for most deep learning frameworks. Are we on the same page or am I missing something?

haydnv replied:
Yes, this sounds fine

haydnv commented on client/tinychain/ml/optimizer.py:
```diff
@@ -13,29 +14,54 @@ class Optimizer(Map):
 class GradientDescent(Optimizer):
     @classmethod
-    def create(cls, learning_rate=F32(0.01)):
+    def create(cls, lr=F32(0.01)):
```
nit: I think `learning_rate` is a more descriptive name, but if you think `lr` makes more sense then you should update the docstring also

haydnv commented on client/tinychain/ml/optimizer.py:
```diff
@@ -13,29 +14,54 @@ class Optimizer(Map):
+
+    @property
+    def beta1(self) -> F32:
+        return self['beta1']
```
You should be able to access `self['beta1']` from within the instance methods of the `Adam` class with the type information preserved (as of commit 83ec5c0), so I don't think you need these `@property`s

haydnv commented on client/tinychain/ml/optimizer.py:
```diff
@@ -13,29 +14,54 @@ class Optimizer(Map):
-                return [update(gradient[n], deltas[n]) for n in range(len(shape))]
-            else:
-                return gradient.write(gradient - (deltas * lr))
+        if i == 1:
```
Two things:
 1. This `if` is only going to run at compile-time but `i` is only set at run-time, so consider `tc.If` instead
 2. I think you actually should initialize `m` and `v` in the `Adam.create` classmethod, and then you won't need any conditional logic here.

haydnv commented on client/tinychain/ml/optimizer.py:
```diff
@@ -45,12 +71,13 @@ def train(model, optimizer, inputs, cost, max_iterations):
     def while_cond(i: UInt, loss: Dense):
         fit = (loss > EPS).any()
-        return fit.logical_and(i < max_iterations)
+        return fit.logical_and(i < max_iterations+1)
```
nit: you could also write `i <= max_iterations`

haydnv commented:
Good catch on the backpropagation error--it should be fixed in commit c89b07eb
